### PR TITLE
Distance map, path rendering, new text rendering, new effects

### PR DESCRIPTION
**Pull request by ubitux**

### CODE REVIEW

File: libnopegl/src/distmap.c [hunk 1]
```diff
@@ -0,0 +1,741 @@
...
+    SCALE_INDEX,
+    BEZIER_X_BUF_INDEX,
+    BEZIER_Y_BUF_INDEX,
+    BEZIER_COUNTS_INDEX,
+    BEZIERGROUP_COUNT_INDEX,
+};
+
+static void load_buffers_data(struct distmap *s,
+                              uint8_t *vert_data, size_t vert_offset,
+                              uint8_t *frag_data, size_t frag_offset)
```
Comment (mbouron) [inline libnopegl/src/distmap.c:291]: `vert_offset` and `frag_offset` are unused (the function uses `s->vert_offset` and `s->frag_offset`.

File: libnopegl/src/api.c [hunk 2]
```diff
@@ -196,8 +196,6 @@ void ngli_ctx_reset(struct ngl_ctx *s, int action)
 #if defined(TARGET_ANDROID)
     ngli_android_ctx_reset(&s->android_ctx);
 #endif
-    ngli_atlas_freep(&s->font_atlas); // allocated by the first node text
```
Comment (mbouron) [inline libnopegl/src/api.c:199]: typo in commit desc: futur -> future.

File: libnopegl/src/api.c [hunk 3]
```diff
@@ -196,11 +198,20 @@ void ngli_ctx_reset(struct ngl_ctx *s, int action)
...
     ngli_pgcache_reset(&s->pgcache);
     ngli_gpu_ctx_freep(&s->gpu_ctx);
     ngli_config_reset(&s->config);
 }
 
+void ngli_free_text_builtin_atlas(void *user_arg, void *data)
+{
+    struct text_builtin_atlas *atlas = data;
+    ngli_distmap_freep(&atlas->distmap);
+    memset(atlas->char_map, 0, sizeof(atlas->char_map));
```
Comment (mbouron) [inline libnopegl/src/api.c:211]: Do we really need to memset char_map here since atlas is destroyed anyway ?

File: libnopegl/src/distmap.c (hunk 1 above)
Comment (ubitux) [reply libnopegl/src/distmap.c]: Mmh yeah, removed

File: libnopegl/src/api.c (hunk 2 above)
Comment (ubitux) [reply libnopegl/src/api.c:199]: Fixed, and changed with your suggestion (just removed the comma)

File: libnopegl/src/api.c (hunk 3 above)
Comment (ubitux) [reply libnopegl/src/api.c:211]: No but I'm just being paranoid about dangling references (a `memset` of 256 bytes should not be expensive, especially at this stage)

Comment (mbouron) [reply libnopegl/src/api.c:211]: I'm asking because we don't do that in other part of the codebase for allocated structure.

Comment (ubitux) [reply libnopegl/src/api.c:211]: We usually do it for pointers, but yeah sure I can get rid of it if you think it's not worth it

Comment (mbouron) [reply libnopegl/src/api.c:211]: I think you can drop it, unless we decide to do that everywhere.

File: libnopegl/src/node_text.c [hunk 4]
```diff
@@ -72,6 +72,11 @@ struct pipeline_desc_fg {
     int32_t user_transform_index;
     int32_t color_index;
     int32_t opacity_index;
+    int32_t outline_index;
+    int32_t outline_color_index;
+    int32_t glow_index;
+    int32_t glow_color_index;
+    int32_t blur_index;
```
Comment (mbouron) [inline libnopegl/src/node_text.c:79]: You will need to pack data and use less attributes: and make that work on macOS/openGL:
```
[ERROR] src/backends/gl/program_gl.c:71 program_check_status: could not link shader: ERROR: Implementation limit of 16 MAX_VERTEX_ATTRIBS (e.g., number of generic plus conventional active vec4 attributes) exceeded, shader uses up to vec4 attribute 16.
```
See https://github.com/nope-project/nope.gl/pull/103.

File: libnopegl/src/api.c (hunk 3 above)
Comment (ubitux) [reply libnopegl/src/api.c]: Alright, removed

File: libnopegl/src/text.c [hunk 5]
```diff
@@ -574,7 +586,12 @@ int ngli_text_set_time(struct text *s, double t)
 
             if ((ret = set_transform( s->data_ptrs.transform  + c * 4 * 4, effect_opts->transform_chain,                       target_t)) < 0 ||
                 (ret = set_vec3_value(s->data_ptrs.color      + c * 4,     effect_opts->color_node,      effect_opts->color,   target_t)) < 0 ||
-                (ret = set_f32_value( s->data_ptrs.color      + c * 4 + 3, effect_opts->opacity_node,    effect_opts->opacity, target_t)) < 0)
+                (ret = set_f32_value( s->data_ptrs.color      + c * 4 + 3, effect_opts->opacity_node,    effect_opts->opacity, target_t)) < 0 ||
+                (ret = set_vec3_value(s->data_ptrs.outline    + c * 4,     effect_opts->outline_color_node, effect_opts->outline_color, target_t)) < 0 ||
+                (ret = set_f32_value( s->data_ptrs.outline    + c * 4 + 4, effect_opts->outline_node,       effect_opts->outline,       target_t)) < 0 ||
```
Comment (mbouron) [inline libnopegl/src/text.c:591]: Isn't it c * 4 +3 ?

Comment (ubitux) [reply libnopegl/src/text.c:591]: Oops yeah. Funny that it wasn't spotted by the CI...

File: tests/texteffect.py [hunk 6]
```diff
@@ -138,3 +138,55 @@ def texteffect_chars_space_nospace(cfg: SceneCfg):
...
+    animkf = [
+        ngl.AnimKeyFrameFloat(0, 0),
+        ngl.AnimKeyFrameFloat(1 / 2, 0.2),
+        ngl.AnimKeyFrameFloat(1, 0),
+    ]
+    effects = [ngl.TextEffect(start=0, end=cfg.duration, glow=ngl.AnimatedFloat(animkf), glow_color=(1, 0, 0))]
+    return ngl.Text("G", effects=effects, aspect_ratio=cfg.aspect_ratio)
+
+
+@test_fingerprint(keyframes=10, tolerance=1)
```
Comment (mbouron) [inline tests/texteffect.py:182]: Can you increase the tolerance to 2. It will allow the test to pass on an Nvidia 1070 (Windows) ?

Comment (ubitux) [reply tests/texteffect.py]: Sure, done

File: libnopegl/src/glsl/distmap.frag [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+#define SMALL_OFF(off) (dot(off, off) <= XERR * XERR)
+
+/* Generated by scripts/aberth-init.py */
+#define K3_0 vec2( 0.866025403784439,  0.500000000000000)
+#define K3_1 vec2(-0.866025403784438,  0.500000000000000)
+#define K3_2 vec2(-0.000000000000000, -1.000000000000000)
+
+int aberth_ehrlich_3(out vec3 roots, float a, float b, float c, float d)
+{
+    float r = 1.0; //pow(abs(d / a), 0.2);
```
Comment (mbouron) [inline libnopegl/src/glsl/distmap.frag:152]: missing space `// pow`

File: libnopegl/src/glsl/distmap.frag [hunk 8]
```diff
@@ -0,0 +1,506 @@
...
+        return stitch_2_roots(topology, r.xy, a, b, c);
+    }
+    if (n == 1) return stitch_1_root(topology, r.x, a, b, c);
+    return vec3(LARGE_FLOAT); // cannot return r because it might have been tainted by root_find3()
+}
+
+vec4 get_color(vec2 p)
+{
+    // TODO: can we reduce the number of polynomial to evaluate for a given pixel?
+
```
Comment (mbouron) [inline libnopegl/src/glsl/distmap.frag:368]: spurious newline ?

File: libnopegl/src/glsl/distmap.frag [hunk 9]
```diff
@@ -0,0 +1,506 @@
...
+        if (r.x > r.y) r.xy = r.yx;
+        return stitch_2_roots(topology, r.xy, a, b, c);
+    }
+    if (n == 1) return stitch_1_root(topology, r.x, a, b, c);
+    return vec3(LARGE_FLOAT); // cannot return r because it might have been tainted by root_find3()
+}
+
+vec4 get_color(vec2 p)
+{
+    // TODO: can we reduce the number of polynomial to evaluate for a given pixel?
```
Comment (mbouron) [inline libnopegl/src/glsl/distmap.frag:367]: nit: // -> /* */ ?

File: scripts/aberth-init.py [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+import math
+import sys
+
+n = int(sys.argv[1])
+for k in range(n):
+    angle = 2 * math.pi / n
+    off = math.pi / (2 * n)
+    z = angle * k + off
+    c, s = math.cos(z), math.sin(z)
+    print(f"#define K{n}_{k} vec2({c:18.15f}, {s:18.15f})")
```
Comment (mbouron) [inline scripts/aberth-init.py:12]: Should part of a separate commit ? Can you add a comment on the purpose of this script ?

**Review by mbouron — APPROVED** [review]
Except from my last cosmetics comment, LGTM. Thanks for your hard work, this is a great addition to the project and pave the way to its new direction !

File: libnopegl/src/glsl/distmap.frag (hunk 8 above)
Comment (ubitux) [reply libnopegl/src/glsl/distmap.frag:368]: Not really, it's a comment for the whole double loops. I moved the comment above the function prototype.

File: libnopegl/src/glsl/distmap.frag (hunk 9 above)
Comment (ubitux) [reply libnopegl/src/glsl/distmap.frag:367]: Sure, done

File: scripts/aberth-init.py (hunk 10 above)
Comment (ubitux) [reply scripts/aberth-init.py:16]: There are references to that script in the distmap shader (above each batch of defines). I added a comment in the script itself for cross reference. I think they go together so I believe its addition belongs in this commit.

**Pull request merged**